domain/secretbackend/service/modelsecretbackendservice.go

```diff
@@ -77,12 +81,54 @@ func (s *ModelSecretBackendService) SetModelSecretBackend(ctx context.Context, b
 		default:
 			// Should never happen.
 			return errors.Errorf("setting model secret backend for unsupported model type %q for model %q",
-				modelType, s.modelID)
+				modelType, s.modelUUID)
 
 		}
 	}
-	if err := s.st.SetModelSecretBackend(ctx, s.modelID, backendName); err != nil {
-		return errors.Errorf("setting model secret backend for %q: %w", s.modelID, err)
+
+	if err := s.checkBackendCompatibility(ctx, backendName); err != nil {
+		return errors.Errorf("checking backend compatibility for %q: %w", backendName, err)
+	}
+
+	if err := s.st.SetModelSecretBackend(ctx, s.modelUUID, backendName); err != nil {
+		return errors.Errorf("setting model secret backend for %q: %w", s.modelUUID, err)
+	}
+	return nil
+}
+
+// checkBackendCompatibility verifies if the specified secret backend is
+// compatible with the model's target Juju agent version.
+// It checks for built-in backends or applies version-specific logic for other
+// backends, returning an error if incompatible.
+func (s *ModelSecretBackendService) checkBackendCompatibility(ctx context.Context, name string) error {
+	if name == provider.Internal || name == kubernetes.BackendName {
+		// Internal and k8s secret backends are compatible with all versions of Juju, since those are built-in backends
+		return nil
 	}
+	backend, err := s.st.GetSecretBackend(ctx, secretbackend.BackendIdentifier{Name: name})
+	if err != nil {
+		return errors.Errorf("getting secret backend %q: %w", name, err)
+	}
+	if backend.BackendType == vault.BackendType {
+		mountPath, _ := backend.Config[vault.MountPathKey].(string)
+		if mountPath != "" {
+			modelVersion, err := s.versionGetter.GetModelTargetAgentVersion(ctx)
+			if err != nil {
+				return errors.Errorf("getting model agent version for %q: %w", s.modelUUID, err)
+			}
+
+			if modelVersion.Compare(semversion.MustParse("3.6.12")) < 0 {
+				return errors.Errorf(
+					"model agent version should be at least 3.6.12 to support %q for %q secret backend",
+					vault.MountPathKey, vault.BackendType).Add(secretbackenderrors.NotSupported)
+			}
+			if modelVersion.Major == 4 && modelVersion.Compare(semversion.MustParse("4.0.1")) < 0 {
+				return errors.Errorf(
+					"model agent version should be at least 4.0.1 to support %q for %q secret backend",
+					vault.MountPathKey, vault.BackendType).Add(secretbackenderrors.NotSupported)
+			}
```

Comment:
This isn't correct, or at least version checking is never the right solution.

So the way that we should do this is based on the ModelConfig registered version. As 3.6 is only at version 3 and 4.0 is at version 4, we should be able to state the right thing.

If for some reason, we changed the behaviour of this in 3.6.12, then that imo should have bumped the facade number. This puts us in a sticky situation has 4.0 is name squatting 4.